### ISSUE TITLE
[HOLD for payment 2023-09-11] [HOLD for payment 2023-09-08] [$1000]  CTRL+K, CTRL+SHIFT+K and CTRL+K again closes everything

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#25527]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Open the app
2. Click CTRL+K, it should open search
3. Click CTRL+SHIFT+K, it should open new group search
4. Again click CTRL+K, it closes everything instead of opening search
## Expected Result: 
App should reopen search page when we again click CTRL+K after CTRL+K, CTRL+SHIFT+K
## Actual Result: 
App closes everything when we again click CTRL+K after CTRL+K, CTRL+SHIFT+K
## Workaround:
Can the user still use Expensify without this being fixed? Have you informed them of the workaround?

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [x] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.3.55-3
**Reproducible in staging?:** y
**Reproducible in production?:**n
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation

https://github.com/Expensify/App/assets/43996225/32c4d539-5f40-4327-b64f-ab5d7dec5a79

https://github.com/Expensify/App/assets/43996225/8ff0059a-deae-447c-b06a-af35dcb79db7

**Expensify/Expensify Issue URL:**
**Issue reported by:** @dhanashree-sawant
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1692434730707179

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01e2284d0efa12310e</li>
        <li>Upwork Job ID: 1693613839544455168</li>
        <li>Last Price Increase: 2023-08-21</li>
<li>Automatic offers: </li>
<ul>
<li>Ollyws | Reviewer | 26306881</li>
<li>shubham1206agra | Contributor | 26306882</li>
<li>dhanashree-sawant | Reporter | 26306883</li>
</ul></li>
    </ul>
</details>

**Comment by pradeepmdk** [comment]
### Proposal
### Please re-state the problem that we are trying to solve in this issue.
CTRL+K, CTRL+SHIFT+K and CTRL+K again closes everything

### What is the root cause of that problem?
https://github.com/Expensify/App/blob/58dc0c9576f8c3d3fa8924a7fec0889232c4b537/src/libs/Navigation/linkTo.js?#L62-L74
here on the right side navigate present `NewGroup_Root` and `Search_Root` both are in different stack navigators once is first time opened it will push when second-time shortcut key pressed state not updating so again pressed whatever in the state its closed.
![Screenshot 2023-08-19 at 8 38 07 PM](https://github.com/Expensify/App/assets/16595846/71fa57cb-ab9b-4a52-8540-f6b8b625fa72)


### What changes do you think we should make in order to solve the problem?
```javascript
return Navigation.navigate(ROUTES.SEARCH, CONST.NAVIGATION.TYPE.UP);
return Navigation.navigate(ROUTES.NEW_GROUP, CONST.NAVIGATION.TYPE.UP);
``` 
in up
`action.type = 'REPLACE';`
replace the route.

**Comment by ShogunFire** [comment]
I don't reproduce 

https://github.com/Expensify/App/assets/19537677/cab36df1-51eb-4389-8e4f-df0f0899aec0

**Comment by pradeepmdk** [comment]
@ShogunFire I am not sure why you are not able to reproduce. but this is still there even though staging is able to reproduce and I checked in locally with the latest code.
https://github.com/Expensify/App/assets/16595846/37aea6ef-4b4b-4937-9dc8-37a2ca56c7c7

**Comment by dhanashree-sawant** [comment]
Yes still able to reproduce

**Comment by mountiny** [comment]
@pradeepmdk This is a deploy blocker, can you point to a change which introduced this behaviour? I dont see it in your proposal and without it its hard to judge as we want to know the root cause of why this is happening in staging but not in production

**Comment by pradeepmdk** [comment]
@mountiny I think this one I saw removed the `replace` https://github.com/Expensify/App/pull/22355/files

**Comment by shubham1206agra** [comment]
@mountiny Please revert https://github.com/Expensify/App/pull/24165
Reverting changes fixes the behaviour
Should I open a PR to revert this?

**Comment by shubham1206agra** [comment]
> @mountiny Please revert #24165 Reverting changes fixes the behaviour Should I open a PR to revert this?

Meanwhile, I am investigating the problem with the current code.

**Comment by mountiny** [comment]
I have created a revert here https://github.com/Expensify/App/pull/25572 could we make sure this bug is not repro there?

**Comment by shubham1206agra** [comment]
Here is the recording 

https://github.com/Expensify/App/assets/58412969/0faa6ab3-ecd8-4c8b-92d1-a90a437b6aee

**Comment by shubham1206agra** [comment]
@mountiny @aimane-chnaif Maybe I have fixed the issue without revert

**Comment by shubham1206agra** [comment]
### Proposal
### Please re-state the problem that we are trying to solve in this issue.

CTRL+K, CTRL+SHIFT+K and CTRL+K again closes everything

### What is the root cause of that problem?

This is happening due to multiple race conditions between `window.history` methods and the onStateChange method execution. Due to this, multiple `window.history.go(-1)` calls were made.

### What changes do you think we should make in order to solve the problem?

By patching the `series` function in `useLinking.js` of `@react-navigation/native` cause original function has bugs.

**Comment by shubham1206agra** [comment]
I have already have branch ready for PR
If you want to make a PR, just confirm here. I will put one up immediately.

**Comment by mountiny** [comment]
@shubham1206agra can you provide a proposal with the solution exactly and show its working as expected please? We wont steal your solution if it works but also want to avoid wasting time in case its not good solution.

**Comment by shubham1206agra** [comment]
https://github.com/Expensify/App/assets/58412969/e8a7ddde-6062-44a7-8e09-711395d12ed8

**Comment by pradeepmdk** [comment]
@mountiny @ShogunFire is correct this pr only causing this issue 
in the previous `historyDelta`  is 0 means this will trigger history.replace({
          path,
          state
         }); 
but now
they are checking the state and removing it from history and directly redirecting to the root one. by using
`await history.go(-staleHistoryDiff);`
<img width="928" alt="Screenshot 2023-08-21 at 7 16 20 PM" src="https://github.com/Expensify/App/assets/16595846/b37c10d1-8290-4906-80e6-c9231cc16385">
that is why we are redirected to root one.

**Comment by shubham1206agra** [comment]
Issue https://github.com/Expensify/App/issues/20908

https://github.com/Expensify/App/assets/58412969/41d8c255-05f3-4701-ab17-26a8cb548e79

**Comment by shubham1206agra** [comment]
Issue https://github.com/Expensify/App/issues/21356

https://github.com/Expensify/App/assets/58412969/99098faa-2f11-4aa8-a6ba-58ac6f18150e

**Comment by shubham1206agra** [comment]
Issue https://github.com/Expensify/App/issues/23796

https://github.com/Expensify/App/assets/58412969/f148a25b-c3e3-49ec-941c-d4f410e0a81f

**Comment by shubham1206agra** [comment]
Does not fix https://github.com/Expensify/App/issues/25412

**Comment by shubham1206agra** [comment]
> Does not fix #25412

Temp logs
```
{
    "index": 1,
    "indexGetter": 1,
    "items": [
        {
            "path": "/r/1705808921226559",
            "id": "8hugcF4Du91dOHzCvk6-l",
            "state": "stack-bteljlS4gW7AcWp49NQpC"
        },
        {
            "selected": "<<<<<<<",
            "path": "/settings",
            "id": "OGJAC-QkjFmJLEzjNjci6",
            "state": "stack-bteljlS4gW7AcWp49NQpC"
        }
    ]
}
createMemoryHistory.js:7 {
    "index": 2,
    "indexGetter": 2,
    "items": [
        {
            "path": "/r/1705808921226559",
            "id": "8hugcF4Du91dOHzCvk6-l",
            "state": "stack-bteljlS4gW7AcWp49NQpC"
        },
        {
            "path": "/settings",
            "id": "OGJAC-QkjFmJLEzjNjci6",
            "state": "stack-bteljlS4gW7AcWp49NQpC"
        },
        {
            "selected": "<<<<<<<",
            "path": "/settings/about",
            "id": "dVlBM7I6DvSCwR2sv5eeY",
            "state": "stack-bteljlS4gW7AcWp49NQpC"
        }
    ]
}
createMemoryHistory.js:7 {
    "index": 3,
    "indexGetter": 3,
    "items": [
        {
            "path": "/r/1705808921226559",
            "id": "8hugcF4Du91dOHzCvk6-l",
            "state": "stack-bteljlS4gW7AcWp49NQpC"
        },
        {
            "path": "/settings",
            "id": "OGJAC-QkjFmJLEzjNjci6",
            "state": "stack-bteljlS4gW7AcWp49NQpC"
        },
        {
            "path": "/settings/about",
            "id": "dVlBM7I6DvSCwR2sv5eeY",
            "state": "stack-bteljlS4gW7AcWp49NQpC"
        },
        {
            "selected": "<<<<<<<",
            "path": "/r/7323501660300024",
            "id": "GhiTr-GpHH6SW8lA25iqS",
            "state": "stack-bteljlS4gW7AcWp49NQpC"
        }
    ]
}
createMemoryHistory.js:7 {
    "index": 3,
    "indexGetter": 3,
    "items": [
        {
            "path": "/r/1705808921226559",
            "id": "8hugcF4Du91dOHzCvk6-l",
            "state": "stack-bteljlS4gW7AcWp49NQpC"
        },
        {
            "path": "/settings",
            "id": "OGJAC-QkjFmJLEzjNjci6",
            "state": "stack-bteljlS4gW7AcWp49NQpC"
        },
        {
            "path": "/settings/about",
            "id": "dVlBM7I6DvSCwR2sv5eeY",
            "state": "stack-bteljlS4gW7AcWp49NQpC"
        },
        {
            "selected": "<<<<<<<",
            "path": "/r/7323501660300024",
            "id": "GhiTr-GpHH6SW8lA25iqS",
            "state": "stack-bteljlS4gW7AcWp49NQpC"
        }
    ]
}
createMemoryHistory.js:7 {
    "index": 2,
    "indexGetter": 2,
    "items": [
        {
            "path": "/r/1705808921226559",
            "id": "8hugcF4Du91dOHzCvk6-l",
            "state": "stack-bteljlS4gW7AcWp49NQpC"
        },
        {
            "path": "/settings",
            "id": "OGJAC-QkjFmJLEzjNjci6",
            "state": "stack-bteljlS4gW7AcWp49NQpC"
        },
        {
            "selected": "<<<<<<<",
            "path": "/settings/about",
            "id": "dVlBM7I6DvSCwR2sv5eeY",
            "state": "stack-bteljlS4gW7AcWp49NQpC"
        },
        {
            "path": "/r/7323501660300024",
            "id": "GhiTr-GpHH6SW8lA25iqS",
            "state": "stack-bteljlS4gW7AcWp49NQpC"
        }
    ]
}
createMemoryHistory.js:7 {
    "index": 2,
    "indexGetter": 2,
    "items": [
        {
            "path": "/r/1705808921226559",
            "id": "8hugcF4Du91dOHzCvk6-l",
            "state": "stack-bteljlS4gW7AcWp49NQpC"
        },
        {
            "path": "/settings",
            "id": "OGJAC-QkjFmJLEzjNjci6",
            "state": "stack-bteljlS4gW7AcWp49NQpC"
        },
        {
            "selected": "<<<<<<<",
            "path": "/settings/about",
            "id": "dVlBM7I6DvSCwR2sv5eeY",
            "state": "stack-bteljlS4gW7AcWp49NQpC"
        },
        {
            "path": "/r/7323501660300024",
            "id": "GhiTr-GpHH6SW8lA25iqS",
            "state": "stack-bteljlS4gW7AcWp49NQpC"
        }
    ]
}
createMemoryHistory.js:7 {
    "index": 1,
    "indexGetter": 1,
    "items": [
        {
            "path": "/r/1705808921226559",
            "id": "8hugcF4Du91dOHzCvk6-l",
            "state": "stack-bteljlS4gW7AcWp49NQpC"
        },
        {
            "selected": "<<<<<<<",
            "path": "/settings",
            "id": "Ddpn6LV45RYRFw_ng7kjD",
            "state": "stack-bteljlS4gW7AcWp49NQpC"
        }
    ]
}
```

**Comment by b4s36t4** [comment]
@shubham1206agra What is the change you've done in patch? the branch includes already existing patch code as well.

**Comment by mountiny** [comment]
> I just changed the series function in useLinking.js as the old implementation has bugs and does not work as intended.

Can you please specify this step a bit more, it will help @adamgrzybowski to review the changes please

**Comment by shubham1206agra** [comment]
I just used this function
```js
const series = (cb) => {
  let queue = Promise.resolve();
  const callback = () => {
    queue = queue.then(cb);
  };
  return callback;
};
```
Which chains the promise to execute sequentially, which the previous function somehow fails to prevent.

**Comment by pradeepmdk** [comment]
@shubham1206agra i saw In the code node_modules currently we have this only after the patch apply

<img width="836" alt="Screenshot 2023-08-21 at 10 38 57 PM" src="https://github.com/Expensify/App/assets/16595846/d1dd7f6c-b316-409a-8a50-ba244a018c60">

how this will resolve this issue


> I just used this function
> 
> ```js
> const series = (cb) => {
>   let queue = Promise.resolve();
>   const callback = () => {
>     queue = queue.then(cb);
>   };
>   return callback;
> };
> ```
> 
> Which chains the promise to execute sequentially, which the previous function somehow fails to prevent.

**Comment by shubham1206agra** [comment]
> @shubham1206agra i saw In the code node_modules currently we have this only after the patch apply
> <img alt="Screenshot 2023-08-21 at 10 38 57 PM" width="836" src="https://user-images.githubusercontent.com/16595846/262095124-d1dd7f6c-b316-409a-8a50-ba244a018c60.png">
> 
> how this will resolve this issue
> 
> > I just used this function
> > ```js
> > const series = (cb) => {
> >   let queue = Promise.resolve();
> >   const callback = () => {
> >     queue = queue.then(cb);
> >   };
> >   return callback;
> > };
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Which chains the promise to execute sequentially, which the previous function somehow fails to prevent.
> 
> I think you are misunderstanding the patch file here `+` or `-` is there right. that only is the changes of patch file
> <img alt="Screenshot 2023-08-21 at 10 44 08 PM" width="715" src="https://user-images.githubusercontent.com/16595846/262096204-d6aec437-9476-41cb-92d3-6fb07e3cd05a.png">
> 
> so in the current change of patch file nothing modifed in the `series` functions

I haven't put the updated patch in main repo.

**Comment by b4s36t4** [comment]
<img width="578" alt="Screenshot 2023-08-22 at 1 18 44 AM" src="https://github.com/Expensify/App/assets/59088937/a1a78bca-0b4c-469b-b0ef-18b4742f3560">

Adding `index` instead of `n` in the and patching it would solve half of the issue. It's just brute force method.

**Comment by adamgrzybowski** [comment]
@shubham1206agra So is your solution the same as the one presented here? https://github.com/react-navigation/react-navigation/pull/11485. That would make sense.

**Comment by shubham1206agra** [comment]
> @shubham1206agra So is your solution the same as the one presented here? https://github.com/react-navigation/react-navigation/pull/11485. That would make sense.

Yes

**Comment by shubham1206agra** [comment]
> @shubham1206agra So is your solution the same as the one presented here? [react-navigation/react-navigation#11485](https://github.com/react-navigation/react-navigation/pull/11485). That would make sense.

And I have applied one extra try-catch block which was missed earlier
And I have returned Promise.resolve() instead of empty return in go method

**Comment by shubham1206agra** [comment]
@adamgrzybowski About #25412 
I don't know what is the ideal behaviour
Can you confirm it?
Cause UI is getting messed up in this

**Comment by shubham1206agra** [comment]
@adamgrzybowski And should I remove log method which was left earlier in `createMemoryHistory.js`

**Comment by adamgrzybowski** [comment]
@shubham1206agra 
We should keep the log function but comment out lines where it is used

Not sure what about extra `try-catch` and `Promise.resolve`. What problem does it solve? Could you please elaborate on it more?

About #25412 we are looking into this one with @WoLewicki. This one is weeeeird 😄

**Comment by b4s36t4** [comment]
> About https://github.com/Expensify/App/issues/25412 we are looking into this one with @WoLewicki. This one is weeeeird 😄

This one won't happen if we add `Navigation.dismissModal()` before going to concierge in `AboutPage.js`

**Comment by DylanDylann** [comment]
https://expensify.slack.com/archives/C049HHMV9SM/p1691856719118529
https://expensify.slack.com/archives/C049HHMV9SM/p1691838672213229

I see 2 above bugs that maybe relate to this issue. @mountiny @Ollyws  Could you check and decide if it could be fixed in this issue

**Comment by shubham1206agra** [comment]
https://expensify.slack.com/archives/C049HHMV9SM/p1691856719118529
This one is expected as Modal.close() maybe was called in the opening of KeyboardShortcutModal. So it closes the YearPickerModal

**Comment by shubham1206agra** [comment]
https://expensify.slack.com/archives/C049HHMV9SM/p1691838672213229
This one I am able to repro

**Comment by shubham1206agra** [comment]
Not sure what about extra try-catch and Promise.resolve. What problem does it solve? Could you please elaborate on it more?
`Promise.resolve` I used to make function behaviour on await more consistent. And try-catch was there so that if somehow await go was called with reject. It will not run the history.push does not run afterwards

**Comment by misgana96** [comment]
@mountiny I think this report is related to this one https://expensify.slack.com/archives/C049HHMV9SM/p1691838672213229.
I reported this issue earlier in the slack

**Comment by mountiny** [comment]
Making this weekly we are looking into getting https://github.com/react-navigation/react-navigation/pull/11485 merged to react navigation upstream for now

**Comment by shubham1206agra** [comment]
Looks like this is the reason behind revert failing
https://github.com/Expensify/App/pull/22437#issuecomment-1691525117

**Comment by shubham1206agra** [comment]
> Although did we figure out the root cause in our code? Because this has worked before right? And the bug in react-navigation is there for a while now

Yes, it is happening due to the patch 
Explained in this comment clearly which part causing the issue
https://github.com/Expensify/App/issues/25527#issuecomment-1686368464

And somewhat explained in this comment
https://github.com/Expensify/App/pull/25572#issuecomment-1686124384

**Comment by mountiny** [comment]
@shubham1206agra thanks! too many comments these days

Alright @Ollyws @adamgrzybowski do you think we should wait for upstream fix or update this in a patch to react naivgation now?

**Comment by adamgrzybowski** [comment]
I would say update the patch. We don't know when the upstream will be ready for `6.x` and we need to bump the version anyway. But I would patch only the `series` function with the code from [this PR ](https://github.com/react-navigation/react-navigation/pull/11485)and nothing else.

**Comment by mountiny** [comment]
Ok ok @shubham1206agra Do you want to do that?

**Comment by shubham1206agra** [comment]
Yes I will do that

**Comment by mountiny** [comment]
@shubham1206agra please go ahead and rasie the PR, thanks!

**Comment by shubham1206agra** [comment]
PR raised

**Comment by mountiny** [comment]
@Ollyws You can do the checklisst

**Comment by shubham1206agra** [comment]
I have raised another PR cause the previous PR got reverted. Should be merged after the build fix

**Comment by tjferriss** [comment]
Hold on, Melvin. We're waiting for this PR to be merged and deployed before we process payment https://github.com/Expensify/App/pull/25989

**Comment by mountiny** [comment]
@Ollyws @shubham1206agra this is not repro on main probably after the styling refactor

However there is a new issue with this switching:
1. Open the Search
2. Use CMD + SHIFT +K to open the new group
3. Use CMD + K  to go back to search

It closes the RHP

or if you do:
1. Open settings
2. CMD + K to open search
3. CMK + SHIFT + K to open new group
4. CMD + K again to open search 
5. The Settings page shows

It seems like when you go there and back it also goeas one back in the RHP stack ie closes it if there is nothign

@shubham1206agra Could you check if your additional patch will fix these? If not would you be interested to investigate these aforementioned issues as part of this issue?

**Comment by shubham1206agra** [comment]
@mountiny I am able to repro the original issue on the latest main. Can you check this again, please?
I have done also `npm ci` after pulling main

**Comment by mountiny** [comment]
I cannot reproduce, but can repro what I wrote [here](https://github.com/Expensify/App/issues/25527#issuecomment-1700391420)
https://github.com/Expensify/App/assets/36083550/33b77874-bb01-4b66-8e4f-83909809bcf5

@shubham1206agra

**Comment by mountiny** [comment]
Perfect, thanks! could you update the Test and QA steps in the PR to follow what was described above and ask the C+ for new review and retest please? cc @Ollyws thank you!

**Comment by mountiny** [comment]
I think we can give th ebonus as the delay was not your fault

**Comment by tjferriss** [comment]
@Ollyws I think you still need to accept the offer in Upworks. Can you also get started on the checklist?

@shubham1206agra and @dhanashree-sawant payments have been sent.

**Comment by shubham1206agra** [comment]
> I think we can give th ebonus as the delay was not your fault

@tjferriss Can you confirm about bonus here?

**Comment by shubham1206agra** [comment]
@tjferriss Can you confirm this?

**Comment by tjferriss** [comment]
Confirmed, yes. I've updated both of your payments to include the bonus amount.

Payment summary:
- @Ollyws Reviewer $1500
- @shubham1206agra Contributor $1500
- @dhanashree-sawant Reporter $250

**Comment by tjferriss** [comment]
Thanks for the detail. I think we're all good here.
